src/mame/konami/konami_gn676_lan.cpp

```diff
@@ -65,176 +89,287 @@ This pcb is the same as the A version but with one added chip:
+		m_fpga_receiving = true;
+	}
 
-			return (r) << 24;
-		}
+	LOG("%s: lanc1_r: read %02X from %08X\n", tag(), r, offset, machine().describe_context());
 
-		default:
+	return r;
+}
+
+void konami_gn676_lan_device::lanc1_w(offs_t offset, uint8_t data)
+{
+	LOG("%s: lanc1_w: wrote %02X to %08X\n", machine().describe_context(), data, offset);
+
+	if (offset == 0x20)
+	{
+		if (BIT(data, 7))
 		{
-			LOG("lanc1_r: %08X at %08X\n", offset, machine().describe_context());
-			return 0xffffffff;
+			m_fpga_uploaded = true;
+			m_fpga_waiting_firmware = false;
+			m_fpga_receiving = false;
+			m_fpga_firmware_crc = util::crc32_creator::simple(m_lanc2_ram.get(), m_fpga_firmware_size);
+
+			LOG("%s: lanc1_fpga: Found firmware with hash of %08x\n", tag(), m_fpga_firmware_crc, machine().describe_context());
+
+			m_x76f041_enabled = m_fpga_firmware_crc == 0xa8c97a75 || m_fpga_firmware_crc == 0xdfc74cc9 || m_fpga_firmware_crc == 0x93b86e35;
+			m_fpga_is_stubbed = m_fpga_firmware_crc == 0x3760e3ce;
+
+			/*
+			// DEBUG: Dump firmware for quick comparison
+			FILE *f = fopen(util::string_format("firmware_%08x.bin", m_fpga_firmware_crc).c_str(), "wb");
+			if (f)
+			{
+				fwrite(m_lanc2_ram.get(), 1, m_fpga_firmware_size, f);
+				fclose(f);
+			}
+			*/
+
+			std::fill_n(m_lanc2_ram.get(), 0x8000, 0);
 		}
 	}
 }
 
-void konami_gn676_lan_device::lanc1_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+uint8_t konami_gn676_lan_device::lanc2_r(offs_t offset)
 {
-	LOG("lanc1_w: %08X, %08X, %08X at %08X\n", data, offset, mem_mask, machine().describe_context());
-}
+	if (m_fpga_waiting_firmware)
+	{
+		if (offset == 0)
+			return 0x80; // Required to be set for firmware to be written
+		return 0;
+	}
 
-uint32_t konami_gn676_lan_device::lanc2_r(offs_t offset, uint32_t mem_mask)
-{
-	uint32_t r = 0;
+	uint8_t r = m_lanc2_reg[offset];
+	if (offset == 1)
+	{
+		r = BIT(m_lanc2_ram_r, 8, 8);
+	}
+	else if (offset == 2)
+	{
+		r = BIT(m_lanc2_ram_r, 0, 8);
+	}
+	else if (offset == 3)
+	{
+		r = m_lanc2_ram[m_lanc2_ram_r];
+		m_lanc2_ram_r = (m_lanc2_ram_r + 1) & 0x7fff;
 
-	if (offset == 0)
+		if (!m_fpga_is_stubbed)
+			r = 0xff; // HACK: network traffic isn't implemented so give dummy data
+	}
+	else if (offset == 6 || offset == 7)
 	{
-		if (ACCESSING_BITS_0_7)
+		if (m_fpga_is_stubbed)
 		{
-			r |= m_lanc2_ram[m_lanc2_ram_r & 0x7fff];
-			m_lanc2_ram_r++;
+			// Thrill Drive's boot test checks for this specific logic
+			r = (m_lanc2_reg[offset] + 3) & 0x7f;
+			m_lanc2_reg[offset]++;
 		}
 		else
-		{
-			r |= 0xffffff00;
-		}
+			r = 0x55; // HACK: force clients to show as disconnected
 	}
-	else if (offset == 1)
+	else if (offset == 9)
 	{
-		r |= 0x00005555;        // set all other machines as disconnected
+		// Thrill Drive's boot test checks for this specific logic
+		uint8_t a = m_lanc2_reg[offset] + 6;
+		uint8_t b = BIT(a, 4, 4) & BIT(a, 5, 3) & BIT(a, 6, 2) & 3;
+		r = (b << 4) | (a & 0x0f);
```

Comment:
When you make the value for `b`, you’re extracting fields wider than two bits, but you have `& 3` at the end, so only the bottom two bits ever make it through.  This doesn’t really make sense.

Reply:
I removed the `& 3` at least but let me work through the math a little before I try simplifying it further. I'm pretty sure I get what you're saying though, just don't want to break it.

Reply:
Ok I'm satisfied with it now, and it's still booting. I'll remove the flag from the systems now (just saw the latest comment, sorry).